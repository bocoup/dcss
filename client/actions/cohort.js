import {
  // CREATE_COHORT,
  CREATE_COHORT_SUCCESS,
  CREATE_COHORT_ERROR,
  SET_COHORT_SUCCESS,
  SET_COHORT_ERROR,
  // GET_COHORT,
  GET_COHORT_SUCCESS,
  GET_COHORT_ERROR,
  // GET_COHORT_PARTICIPANTS,
  GET_COHORT_PARTICIPANTS_SUCCESS,
  GET_COHORT_PARTICIPANTS_ERROR,
  // GET_COHORT_RUN_DATA_SUCCESS,
  GET_COHORT_RUN_DATA_ERROR,
  // GET_ALL_COHORTS,
  GET_ALL_COHORTS_SUCCESS,
  GET_ALL_COHORTS_ERROR,
  // GET_USER_COHORTS,
  GET_USER_COHORTS_SUCCESS,
  GET_USER_COHORTS_ERROR,
  // LINK_COHORT_TO_RUN,
  LINK_RUN_TO_COHORT_SUCCESS,
  LINK_RUN_TO_COHORT_ERROR,
  // SET_COHORT_USER_ROLE,
  SET_COHORT_USER_ROLE_SUCCESS,
  SET_COHORT_USER_ROLE_ERROR
} from './types';

export const createCohort = ({ name }) => async dispatch => {
  try {
    const res = await (await fetch('/api/cohort', {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ name })
    })).json();

    if (res.error) {
      throw res;
    }
    const { cohort } = res;

    dispatch({ type: CREATE_COHORT_SUCCESS, cohort });
    // return the cohort to the promise action for redirection purposes
    return cohort;
  } catch (error) {
    dispatch({ type: CREATE_COHORT_ERROR, error });
    return null;
  }
};

export const setCohort = cohort => async dispatch => {
  // dispatch({ type: SET_COHORT, cohort });
  try {
    const { scenarios } = cohort;
    const body = JSON.stringify({
      scenarios
    });
    await (await fetch(`/api/cohort/${cohort.id}/scenarios`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body
    })).json();
    dispatch({ type: SET_COHORT_SUCCESS, cohort });
  } catch (error) {
    dispatch({ type: SET_COHORT_ERROR, error });
    return null;
  }
};

export const getCohort = id => async dispatch => {
  if (Number.isNaN(id)) {
    return;
  }
  try {
    const res = await (await fetch(`/api/cohort/${id}`)).json();
    if (res.error) {
      throw res;
    }
    const { cohort } = res;
    dispatch({ type: GET_COHORT_SUCCESS, cohort });
    // return the cohort to the promise action for redirection purposes
    return cohort;
  } catch (error) {
    dispatch({ type: GET_COHORT_ERROR, error });
    return null;
  }
};

export const getCohorts = () => async dispatch => {
  try {
    const response = await (await fetch('/api/cohort/my')).json();

    if (response.error) {
      throw response;
    }
    const { cohorts } = response;
    dispatch({ type: GET_USER_COHORTS_SUCCESS, cohorts });
  } catch (error) {
    dispatch({ type: GET_USER_COHORTS_ERROR, error });
    return null;
  }
};

export const getAllCohorts = () => async dispatch => {
  try {
    const res = await (await fetch('/api/cohort/all')).json();
    if (res.error) {
      throw res;
    }
    const { cohorts } = res;
    dispatch({ type: GET_ALL_COHORTS_SUCCESS, cohorts });
    return cohorts;
  } catch (error) {
    dispatch({ type: GET_ALL_COHORTS_ERROR, error });
    return null;
  }
};

export const getCohortParticipants = id => async dispatch => {
  try {
    const res = await (await fetch(`/api/cohort/${id}`)).json();

    if (res.error) {
      throw res;
    }
    const {
      cohort,
      cohort: { users }
    } = res;
    // Dispatch the entire "cohort", but only the "users" property
    // will be used in the reducer.
    dispatch({ type: GET_COHORT_PARTICIPANTS_SUCCESS, cohort });
    // return the cohort to the promise action for redirection purposes
    return users;
  } catch (error) {
    dispatch({ type: GET_COHORT_PARTICIPANTS_ERROR, error });
    return null;
  }
};

export const getCohortData = (
  cohort_id,
  participant_id,
  scenario_id
) => async dispatch => {
  try {
    const endpoint = scenario_id
      ? `/api/cohort/${cohort_id}/scenario/${scenario_id}`
      : `/api/cohort/${cohort_id}/participant/${participant_id}`;

    const res = await (await fetch(endpoint)).json();

    if (res.error) {
      throw res;
    }

    const { prompts, responses } = res;

    // This dispatch does not update internal state, since
    // such state would be thrown away:
    // dispatch({ type: GET_COHORT_RUN_DATA_SUCCESS, prompts, responses });
    //
    return { prompts, responses };
  } catch (error) {
    dispatch({ type: GET_COHORT_RUN_DATA_ERROR, error });
    return null;
  }
};

export const linkRunToCohort = (cohort_id, run_id) => async dispatch => {
  try {
    const res = await (await fetch(
      `/api/cohort/${cohort_id}/run/${run_id}`
    )).json();
    if (res.error) {
      throw res;
    }
    const { cohort } = res;
    dispatch({ type: LINK_RUN_TO_COHORT_SUCCESS });
    dispatch({ type: GET_COHORT_SUCCESS, cohort });
    return cohort;
  } catch (error) {
    dispatch({ type: LINK_RUN_TO_COHORT_ERROR, error });
    return null;
  }
};
// This is used to
export const linkUserToCohort = (cohort_id, role) => async dispatch => {
  try {
    const users = await (await fetch(
      `/api/cohort/${cohort_id}/join/${role}`
    )).json();
    if (users.error) {
      throw users;
    }
    dispatch({ type: SET_COHORT_USER_ROLE_SUCCESS, users });
  } catch (error) {
    dispatch({ type: SET_COHORT_USER_ROLE_ERROR, error });
    return null;
  }
};

export const addCohortUserRole = (
  cohort_id,
  user_id,
  role
) => async dispatch => {
  try {
    const body = JSON.stringify({
      cohort_id,
      user_id,
      roles: [role]
    });
    const result = await (await fetch('/api/cohort/${cohort_id}/roles/add', {
      headers: {
        'Content-Type': 'application/json'
      },
      method: 'POST',
      body
    })).json();

    const { cohort } = result;
    dispatch({ type: SET_COHORT_USER_ROLE_SUCCESS });
    dispatch({ type: GET_COHORT_SUCCESS, cohort });
    return result;
  } catch (error) {
    dispatch({ type: SET_COHORT_USER_ROLE_ERROR, error });
    return null;
  }
};

export const deleteCohortUserRole = (
  cohort_id,
  user_id,
  role
) => async dispatch => {
  try {
    const body = JSON.stringify({
      cohort_id,
      user_id,
      roles: [role]
    });
    const result = await (await fetch('/api/cohort/${cohort_id}/roles/delete', {
      headers: {
        'Content-Type': 'application/json'
      },
      method: 'POST',
      body
    })).json();
    const { cohort } = result;
    dispatch({ type: SET_COHORT_USER_ROLE_SUCCESS });
    dispatch({ type: GET_COHORT_SUCCESS, cohort });
    return result;
  } catch (error) {
    dispatch({ type: SET_COHORT_USER_ROLE_ERROR, error });
    return null;
  }
};
