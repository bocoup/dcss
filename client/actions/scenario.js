import {
    GET_SCENARIO,
    GET_SCENARIO_SUCCESS,
    GET_SCENARIO_ERROR,
    GET_SCENARIOS,
    GET_SCENARIOS_SUCCESS,
    GET_SCENARIOS_ERROR,
    GET_SCENARIO_RUN_HISTORY,
    GET_SCENARIO_RUN_HISTORY_SUCCESS,
    GET_SCENARIO_RUN_HISTORY_ERROR,
    GET_SLIDES,
    GET_SLIDES_SUCCESS,
    GET_SLIDES_ERROR,
    SET_SCENARIO,
    // SET_SCENARIO_SUCCESS,
    // SET_SCENARIO_ERROR,
    SET_SCENARIOS,
    // SET_SCENARIOS_SUCCESS,
    // SET_SCENARIOS_ERROR,
    SET_SLIDES
} from './types';

export const getScenarios = () => async dispatch => {
    dispatch({ type: GET_SCENARIOS });
    try {
        const { scenarios, error } = await (await fetch(
            '/api/scenarios'
        )).json();
        if (error) {
            throw error;
        }

        dispatch({ type: GET_SCENARIOS_SUCCESS, scenarios });
        return scenarios;
    } catch (error) {
        const { message, status, stack } = error;
        dispatch({ type: GET_SCENARIOS_ERROR, status, message, stack });
    }
};

export const getScenario = id => async dispatch => {
    dispatch({ type: GET_SCENARIO, id });
    try {
        const { scenario, error } = await (await fetch(
            `/api/scenarios/${id}`
        )).json();

        if (error) {
            throw error;
        }
        dispatch({ type: GET_SCENARIO_SUCCESS, scenario });
        return scenario;
    } catch (error) {
        const { message, status, stack } = error;
        dispatch({ type: GET_SCENARIO_ERROR, status, message, stack });
    }
};

export const getScenarioRunHistory = params => async dispatch => {
    const { scenarioId, cohortId } = params;
    dispatch({ type: GET_SCENARIO_RUN_HISTORY });
    try {
        const endpoint = cohortId
            ? `/api/scenarios/${scenarioId}/cohort/${cohortId}/history`
            : `/api/scenarios/${scenarioId}/history`;

        const { history, error } = await (await fetch(endpoint)).json();

        if (error) {
            throw error;
        }

        dispatch({ type: GET_SCENARIO_RUN_HISTORY_SUCCESS, ...history });
        return { ...history };
    } catch (error) {
        const { message, stack, status } = error;
        dispatch({
            type: GET_SCENARIO_RUN_HISTORY_ERROR,
            message,
            stack,
            status
        });
        // pass along the error to the promise action
        throw error;
    }
};

export const getSlides = scenarioId => async dispatch => {
    dispatch({ type: GET_SLIDES, scenarioId });
    try {
        const { slides, error } = await (await fetch(
            `/api/scenarios/${scenarioId}/slides`
        )).json();

        if (error) {
            throw error;
        }
        dispatch({ type: GET_SLIDES_SUCCESS, slides });
        return slides;
    } catch (error) {
        const { message, status, stack } = error;
        dispatch({ type: GET_SLIDES_ERROR, status, message, stack });
    }
};

export const setScenario = scenario => ({
    type: SET_SCENARIO,
    scenario
});

export const setScenarios = scenarios => ({
    type: SET_SCENARIOS,
    scenarios
});

export const setSlides = slides => ({
    type: SET_SLIDES,
    slides
});
