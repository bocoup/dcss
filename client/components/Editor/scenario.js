export function makeDefaultDescription({ title, description }) {
  let returnValue = description;
  if (!returnValue && title) {
    returnValue = `A scenario about "${title}"`;
  }
  return returnValue;
}
