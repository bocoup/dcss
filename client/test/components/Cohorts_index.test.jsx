import React from 'react';
jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useLayoutEffect: jest.requireActual('react').useEffect
}));

import assert from 'assert';
import {
  fetchImplementation,
  mounter,
  reduxer,
  serialize,
  snapshotter,
  state
} from '../bootstrap';
import { unmountComponentAtNode } from 'react-dom';

import { mount, shallow } from 'enzyme';
import {
  fireEvent,
  prettyDOM,
  render,
  screen,
  waitFor
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import {
  GET_ALL_COHORTS_SUCCESS,
  GET_USER_COHORTS_SUCCESS,
  GET_COHORT_SUCCESS,
  SET_COHORT_SUCCESS,
  CREATE_COHORT_SUCCESS,
  GET_SCENARIOS_SUCCESS,
  GET_USER_SUCCESS
} from '../../actions/types';
import * as cohortActions from '@actions/cohort';
import * as scenarioActions from '@actions/scenario';
import * as userActions from '@actions/user';
jest.mock('../../actions/cohort');
jest.mock('../../actions/scenario');
jest.mock('../../actions/user');

import Cohorts from '../../components/Cohorts/index.jsx';

const original = JSON.parse(JSON.stringify(state));
let container = null;
let commonProps = null;
let commonState = null;

beforeAll(() => {
  (window || global).fetch = jest.fn();
});

afterAll(() => {
  jest.restoreAllMocks();
});

beforeEach(() => {
  container = document.createElement('div');
  container.setAttribute('id', 'root');
  document.body.appendChild(container);

  fetchImplementation(fetch);

  cohortActions.getAllCohorts = jest.fn();
  cohortActions.getAllCohorts.mockImplementation(() => async dispatch => {
    const cohorts = [
      {
        id: 1,
        created_at: '2020-08-31T14:01:08.656Z',
        name: 'A New Cohort That Exists Within Inline Props',
        runs: [
          {
            id: 11,
            user_id: 333,
            scenario_id: 99,
            created_at: '2020-03-28T19:44:03.069Z',
            updated_at: '2020-03-31T17:01:43.139Z',
            ended_at: '2020-03-31T17:01:43.128Z',
            consent_id: 8,
            consent_acknowledged_by_user: true,
            consent_granted_by_user: true,
            referrer_params: null,
            cohort_id: 1,
            run_id: 11
          }
        ],
        scenarios: [99],
        users: [
          {
            username: 'super',
            personalname: 'Super User',
            email: 'super@email.com',
            id: 999,
            roles: ['participant', 'super_admin'],
            is_anonymous: false,
            is_super: true
          },
          {
            username: 'facilitator',
            personalname: 'Facilitator User',
            email: 'facilitator@email.com',
            id: 555,
            roles: ['participant', 'facilitator', 'researcher'],
            is_anonymous: false,
            is_super: false
          },
          {
            username: 'researcher',
            personalname: 'Researcher User',
            email: 'researcher@email.com',
            id: 444,
            roles: ['participant', 'researcher'],
            is_anonymous: false,
            is_super: false
          },
          {
            username: 'participant',
            personalname: 'Participant User',
            email: 'participant@email.com',
            id: 333,
            roles: ['participant'],
            is_anonymous: false,
            is_super: false
          }
        ],
        roles: ['super', 'facilitator'],
        usersById: {
          999: {
            username: 'super',
            personalname: 'Super User',
            email: 'super@email.com',
            id: 999,
            roles: ['participant', 'super_admin'],
            is_anonymous: false,
            is_super: true
          },
          555: {
            username: 'facilitator',
            personalname: 'Facilitator User',
            email: 'facilitator@email.com',
            id: 555,
            roles: ['participant', 'facilitator', 'researcher'],
            is_anonymous: false,
            is_super: false
          },
          444: {
            username: 'researcher',
            personalname: 'Researcher User',
            email: 'researcher@email.com',
            id: 444,
            roles: ['participant', 'researcher'],
            is_anonymous: false,
            is_super: false
          },
          333: {
            username: 'participant',
            personalname: 'Participant User',
            email: 'participant@email.com',
            id: 333,
            roles: ['participant'],
            is_anonymous: false,
            is_super: false
          }
        }
      }
    ];
    dispatch({ type: GET_ALL_COHORTS_SUCCESS, cohorts });
    return cohorts;
  });
  cohortActions.getCohorts = jest.fn();
  cohortActions.getCohorts.mockImplementation(() => async dispatch => {
    const cohorts = [
      {
        id: 1,
        created_at: '2020-08-31T14:01:08.656Z',
        name: 'A New Cohort That Exists Within Inline Props',
        runs: [
          {
            id: 11,
            user_id: 333,
            scenario_id: 99,
            created_at: '2020-03-28T19:44:03.069Z',
            updated_at: '2020-03-31T17:01:43.139Z',
            ended_at: '2020-03-31T17:01:43.128Z',
            consent_id: 8,
            consent_acknowledged_by_user: true,
            consent_granted_by_user: true,
            referrer_params: null,
            cohort_id: 1,
            run_id: 11
          }
        ],
        scenarios: [99],
        users: [
          {
            username: 'super',
            personalname: 'Super User',
            email: 'super@email.com',
            id: 999,
            roles: ['participant', 'super_admin'],
            is_anonymous: false,
            is_super: true
          },
          {
            username: 'facilitator',
            personalname: 'Facilitator User',
            email: 'facilitator@email.com',
            id: 555,
            roles: ['participant', 'facilitator', 'researcher'],
            is_anonymous: false,
            is_super: false
          },
          {
            username: 'researcher',
            personalname: 'Researcher User',
            email: 'researcher@email.com',
            id: 444,
            roles: ['participant', 'researcher'],
            is_anonymous: false,
            is_super: false
          },
          {
            username: 'participant',
            personalname: 'Participant User',
            email: 'participant@email.com',
            id: 333,
            roles: ['participant'],
            is_anonymous: false,
            is_super: false
          }
        ],
        roles: ['super', 'facilitator'],
        usersById: {
          999: {
            username: 'super',
            personalname: 'Super User',
            email: 'super@email.com',
            id: 999,
            roles: ['participant', 'super_admin'],
            is_anonymous: false,
            is_super: true
          },
          555: {
            username: 'facilitator',
            personalname: 'Facilitator User',
            email: 'facilitator@email.com',
            id: 555,
            roles: ['participant', 'facilitator', 'researcher'],
            is_anonymous: false,
            is_super: false
          },
          444: {
            username: 'researcher',
            personalname: 'Researcher User',
            email: 'researcher@email.com',
            id: 444,
            roles: ['participant', 'researcher'],
            is_anonymous: false,
            is_super: false
          },
          333: {
            username: 'participant',
            personalname: 'Participant User',
            email: 'participant@email.com',
            id: 333,
            roles: ['participant'],
            is_anonymous: false,
            is_super: false
          }
        }
      }
    ];
    dispatch({ type: GET_USER_COHORTS_SUCCESS, cohorts });
    return cohorts;
  });
  cohortActions.getCohort = jest.fn();
  cohortActions.getCohort.mockImplementation(() => async dispatch => {
    const cohort = {
      id: 1,
      created_at: '2020-08-31T14:01:08.656Z',
      name: 'A New Cohort That Exists Within Inline Props',
      runs: [
        {
          id: 11,
          user_id: 333,
          scenario_id: 99,
          created_at: '2020-03-28T19:44:03.069Z',
          updated_at: '2020-03-31T17:01:43.139Z',
          ended_at: '2020-03-31T17:01:43.128Z',
          consent_id: 8,
          consent_acknowledged_by_user: true,
          consent_granted_by_user: true,
          referrer_params: null,
          cohort_id: 1,
          run_id: 11
        }
      ],
      scenarios: [99],
      users: [
        {
          username: 'super',
          personalname: 'Super User',
          email: 'super@email.com',
          id: 999,
          roles: ['participant', 'super_admin'],
          is_anonymous: false,
          is_super: true
        },
        {
          username: 'facilitator',
          personalname: 'Facilitator User',
          email: 'facilitator@email.com',
          id: 555,
          roles: ['participant', 'facilitator', 'researcher'],
          is_anonymous: false,
          is_super: false
        },
        {
          username: 'researcher',
          personalname: 'Researcher User',
          email: 'researcher@email.com',
          id: 444,
          roles: ['participant', 'researcher'],
          is_anonymous: false,
          is_super: false
        },
        {
          username: 'participant',
          personalname: 'Participant User',
          email: 'participant@email.com',
          id: 333,
          roles: ['participant'],
          is_anonymous: false,
          is_super: false
        }
      ],
      roles: ['super', 'facilitator'],
      usersById: {
        999: {
          username: 'super',
          personalname: 'Super User',
          email: 'super@email.com',
          id: 999,
          roles: ['participant', 'super_admin'],
          is_anonymous: false,
          is_super: true
        },
        555: {
          username: 'facilitator',
          personalname: 'Facilitator User',
          email: 'facilitator@email.com',
          id: 555,
          roles: ['participant', 'facilitator', 'researcher'],
          is_anonymous: false,
          is_super: false
        },
        444: {
          username: 'researcher',
          personalname: 'Researcher User',
          email: 'researcher@email.com',
          id: 444,
          roles: ['participant', 'researcher'],
          is_anonymous: false,
          is_super: false
        },
        333: {
          username: 'participant',
          personalname: 'Participant User',
          email: 'participant@email.com',
          id: 333,
          roles: ['participant'],
          is_anonymous: false,
          is_super: false
        }
      }
    };
    dispatch({ type: GET_COHORT_SUCCESS, cohort });
    return cohort;
  });
  cohortActions.setCohort = jest.fn();
  cohortActions.setCohort.mockImplementation(cohort => async dispatch => {
    dispatch({ type: SET_COHORT_SUCCESS, cohort });
    return cohort;
  });
  cohortActions.createCohort = jest.fn();
  cohortActions.createCohort.mockImplementation(cohort => async dispatch => {
    dispatch({ type: CREATE_COHORT_SUCCESS, cohort });
    return cohort;
  });
  scenarioActions.getScenariosByStatus = jest.fn();
  scenarioActions.getScenariosByStatus.mockImplementation(
    () => async dispatch => {
      const scenarios = [
        {
          author: {
            id: 999,
            username: 'super',
            personalname: 'Super User',
            email: 'super@email.com',
            is_anonymous: false,
            roles: ['participant', 'super_admin', 'facilitator', 'researcher'],
            is_super: true
          },
          categories: [],
          consent: { id: 57, prose: '' },
          description: "This is the description of 'A Multiplayer Scenario'",
          finish: {
            id: 1,
            title: '',
            components: [
              { html: '<h2>Thanks for participating!</h2>', type: 'Text' }
            ],
            is_finish: true
          },
          lock: {
            scenario_id: 42,
            user_id: 999,
            created_at: '2020-02-31T23:54:19.934Z',
            ended_at: null
          },
          slides: [
            {
              id: 1,
              title: '',
              components: [
                { html: '<h2>Thanks for participating!</h2>', type: 'Text' }
              ],
              is_finish: true
            },
            {
              id: 2,
              title: '',
              components: [
                {
                  id: 'b7e7a3f1-eb4e-4afa-8569-eb6677358c9e',
                  html: '<p>paragraph</p>',
                  type: 'Text'
                },
                {
                  id: 'aede9380-c7a3-4ef7-add7-838fd5ec854f',
                  type: 'TextResponse',
                  header: 'TextResponse-1',
                  prompt: '',
                  timeout: 0,
                  recallId: '',
                  required: true,
                  responseId: 'be99fe9b-fa0d-4ab7-8541-1bfd1ef0bf11',
                  placeholder: 'Your response'
                },
                {
                  id: 'f96ac6de-ac6b-4e06-bd97-d97e12fe72c1',
                  html: '<p>?</p>',
                  type: 'Text'
                }
              ],
              is_finish: false
            }
          ],
          status: 1,
          title: 'Multiplayer Scenario 2',
          users: [
            {
              id: 999,
              email: 'super@email.com',
              username: 'super',
              personalname: 'Super User',
              roles: ['super'],
              is_super: true,
              is_author: true,
              is_reviewer: false
            }
          ],
          id: 42,
          created_at: '2020-08-31T17:50:28.089Z',
          updated_at: null,
          deleted_at: null
        },
        {
          author: {
            id: 999,
            username: 'super',
            personalname: 'Super User',
            email: 'super@email.com',
            is_anonymous: false,
            roles: ['participant', 'super_admin', 'facilitator', 'researcher'],
            is_super: true
          },
          categories: [],
          consent: { id: 69, prose: '' },
          description: "This is the description of 'Some Other Scenario'",
          finish: {
            id: 11,
            title: '',
            components: [{ html: '<h2>Bye!</h2>', type: 'Text' }],
            is_finish: true
          },
          lock: {
            scenario_id: 42,
            user_id: 999,
            created_at: '2020-02-31T23:54:19.934Z',
            ended_at: null
          },
          slides: [
            {
              id: 11,
              title: '',
              components: [{ html: '<h2>Bye!</h2>', type: 'Text' }],
              is_finish: true
            },
            {
              id: 22,
              title: '',
              components: [
                {
                  id: 'b7e7a3f1-eb4e-4afa-8569-838fd5ec854f',
                  html: '<p>HTML!</p>',
                  type: 'Text'
                },
                {
                  id: 'aede9380-c7a3-4ef7-add7-eb6677358c9e',
                  type: 'TextResponse',
                  header: 'TextResponse-1',
                  prompt: '',
                  timeout: 0,
                  recallId: '',
                  required: true,
                  responseId: 'be99fe9b-fa0d-4ab7-8541-1bfd1ef0bf11',
                  placeholder: 'Your response'
                },
                {
                  id: 'f96ac6de-ac6b-4e06-bd97-d97e12fe72c1',
                  html: '<p>?</p>',
                  type: 'Text'
                }
              ],
              is_finish: false
            }
          ],
          status: 1,
          title: 'Some Other Scenario',
          users: [
            {
              id: 999,
              email: 'super@email.com',
              username: 'super',
              personalname: 'Super User',
              roles: ['super'],
              is_super: true,
              is_author: true,
              is_reviewer: false
            }
          ],
          id: 99,
          created_at: '2020-07-31T17:50:28.089Z',
          updated_at: null,
          deleted_at: null
        }
      ];
      dispatch({ type: GET_SCENARIOS_SUCCESS, scenarios });
      return scenarios;
    }
  );
  userActions.getUser = jest.fn();
  userActions.getUser.mockImplementation(() => async dispatch => {
    const user = {
      username: 'super',
      personalname: 'Super User',
      email: 'super@email.com',
      id: 999,
      roles: ['participant', 'super_admin'],
      is_anonymous: false,
      is_super: true
    };
    dispatch({ type: GET_USER_SUCCESS, user });
    return user;
  });

  commonProps = {};
  commonState = JSON.parse(JSON.stringify(original));
});

afterEach(() => {
  jest.resetAllMocks();
  unmountComponentAtNode(container);
  container.remove();
  container = null;
  commonProps = null;
  commonState = null;
});

test('Cohorts', () => {
  expect(Cohorts).toBeDefined();
});

test('Render 1 1', async done => {
  const Component = Cohorts;

  const props = {
    ...commonProps,
    authority: {}
  };

  const state = {
    ...commonState
  };

  const ConnectedRoutedComponent = reduxer(Component, props, state);

  const { asFragment } = render(<ConnectedRoutedComponent {...props} />);
  expect(asFragment()).toMatchSnapshot();

  done();
});

test('Render 1 2', async done => {
  const Component = Cohorts;

  const props = {
    ...commonProps,
    authority: {}
  };

  const state = {
    session: {
      isLoggedIn: true,
      username: 'super',
      permissions: [
        'create_cohort',
        'edit_scenarios_in_cohort',
        'view_scenarios_in_cohort',
        'edit_own_cohorts',
        'view_all_cohorts',
        'view_own_cohorts',
        'view_all_data',
        'view_consented_data',
        'view_invited_cohorts',
        'view_own_data',
        'edit_all_cohorts',
        'create_scenario',
        'edit_scenario',
        'edit_permissions',
        'view_run_data',
        'view_all_run_data'
      ],
      personalname: 'Super User',
      email: 'super@email.com',
      id: 2,
      roles: ['participant', 'super_admin', 'facilitator', 'researcher'],
      is_anonymous: false,
      is_super: true
    }
  };

  const ConnectedRoutedComponent = reduxer(Component, props, state);

  const { asFragment } = render(<ConnectedRoutedComponent {...props} />);
  expect(asFragment()).toMatchSnapshot();

  done();
});

/*{INJECTION}*/
