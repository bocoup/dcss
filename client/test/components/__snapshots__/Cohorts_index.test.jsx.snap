// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render 1 1 1`] = `"<div class=\\"ui borderless icon em__height menu\\"><a tabindex=\\"0\\" class=\\"item\\"><i class=\\"icons em__icon-group-margin\\"><i aria-hidden=\\"true\\" class=\\"group icon\\"></i><i aria-hidden=\\"true\\" class=\\"green add top right corner icon\\"></i></i>Create a Cohort</a></div><div class=\\"ui fluid container\\"><div class=\\"ui grid\\"><div class=\\"row\\"><div class=\\"stretched column\\"><div class=\\"ui doubling four cards\\"><div style=\\"height: 20rem;\\" class=\\"ui card loading__multi-card\\"><div class=\\"content loading__content\\"><div class=\\"ui placeholder fluid\\"><div class=\\"square image\\"></div></div></div></div><div style=\\"height: 20rem;\\" class=\\"ui card loading__multi-card\\"><div class=\\"content loading__content\\"><div class=\\"ui placeholder fluid\\"><div class=\\"square image\\"></div></div></div></div><div style=\\"height: 20rem;\\" class=\\"ui card loading__multi-card\\"><div class=\\"content loading__content\\"><div class=\\"ui placeholder fluid\\"><div class=\\"square image\\"></div></div></div></div><div style=\\"height: 20rem;\\" class=\\"ui card loading__multi-card\\"><div class=\\"content loading__content\\"><div class=\\"ui placeholder fluid\\"><div class=\\"square image\\"></div></div></div></div><div style=\\"height: 20rem;\\" class=\\"ui card loading__multi-card\\"><div class=\\"content loading__content\\"><div class=\\"ui placeholder fluid\\"><div class=\\"square image\\"></div></div></div></div><div style=\\"height: 20rem;\\" class=\\"ui card loading__multi-card\\"><div class=\\"content loading__content\\"><div class=\\"ui placeholder fluid\\"><div class=\\"square image\\"></div></div></div></div><div style=\\"height: 20rem;\\" class=\\"ui card loading__multi-card\\"><div class=\\"content loading__content\\"><div class=\\"ui placeholder fluid\\"><div class=\\"square image\\"></div></div></div></div><div style=\\"height: 20rem;\\" class=\\"ui card loading__multi-card\\"><div class=\\"content loading__content\\"><div class=\\"ui placeholder fluid\\"><div class=\\"square image\\"></div></div></div></div></div></div></div><div class=\\"row\\"><div class=\\"stretched column\\"></div></div></div></div>"`;

exports[`Render 1 1 2`] = `"<div class=\\"ui borderless icon em__height menu\\"><a tabindex=\\"0\\" class=\\"item\\"><i class=\\"icons em__icon-group-margin\\"><i aria-hidden=\\"true\\" class=\\"group icon\\"></i><i aria-hidden=\\"true\\" class=\\"green add top right corner icon\\"></i></i>Create a Cohort</a></div><div class=\\"ui fluid container\\"><div class=\\"ui grid\\"><div class=\\"row\\"><div class=\\"stretched column\\"><div class=\\"ui doubling four cards\\"><div style=\\"height: 20rem;\\" class=\\"ui card loading__multi-card\\"><div class=\\"content loading__content\\"><div class=\\"ui placeholder fluid\\"><div class=\\"square image\\"></div></div></div></div><div style=\\"height: 20rem;\\" class=\\"ui card loading__multi-card\\"><div class=\\"content loading__content\\"><div class=\\"ui placeholder fluid\\"><div class=\\"square image\\"></div></div></div></div><div style=\\"height: 20rem;\\" class=\\"ui card loading__multi-card\\"><div class=\\"content loading__content\\"><div class=\\"ui placeholder fluid\\"><div class=\\"square image\\"></div></div></div></div><div style=\\"height: 20rem;\\" class=\\"ui card loading__multi-card\\"><div class=\\"content loading__content\\"><div class=\\"ui placeholder fluid\\"><div class=\\"square image\\"></div></div></div></div><div style=\\"height: 20rem;\\" class=\\"ui card loading__multi-card\\"><div class=\\"content loading__content\\"><div class=\\"ui placeholder fluid\\"><div class=\\"square image\\"></div></div></div></div><div style=\\"height: 20rem;\\" class=\\"ui card loading__multi-card\\"><div class=\\"content loading__content\\"><div class=\\"ui placeholder fluid\\"><div class=\\"square image\\"></div></div></div></div><div style=\\"height: 20rem;\\" class=\\"ui card loading__multi-card\\"><div class=\\"content loading__content\\"><div class=\\"ui placeholder fluid\\"><div class=\\"square image\\"></div></div></div></div><div style=\\"height: 20rem;\\" class=\\"ui card loading__multi-card\\"><div class=\\"content loading__content\\"><div class=\\"ui placeholder fluid\\"><div class=\\"square image\\"></div></div></div></div></div></div></div><div class=\\"row\\"><div class=\\"stretched column\\"></div></div></div></div>"`;

exports[`Render 1 1 3`] = `"<div class=\\"ui borderless icon em__height menu\\"><a tabindex=\\"0\\" class=\\"item\\"><i class=\\"icons em__icon-group-margin\\"><i aria-hidden=\\"true\\" class=\\"group icon\\"></i><i aria-hidden=\\"true\\" class=\\"green add top right corner icon\\"></i></i>Create a Cohort</a></div><div class=\\"ui fluid container\\"><div class=\\"ui grid\\"><div class=\\"row\\"><div class=\\"stretched column\\"><div class=\\"ui doubling four cards\\"><div style=\\"height:20rem\\" class=\\"ui card loading__multi-card\\"><div class=\\"content loading__content\\"><div class=\\"ui placeholder fluid\\"><div class=\\"square image\\"></div></div></div></div><div style=\\"height:20rem\\" class=\\"ui card loading__multi-card\\"><div class=\\"content loading__content\\"><div class=\\"ui placeholder fluid\\"><div class=\\"square image\\"></div></div></div></div><div style=\\"height:20rem\\" class=\\"ui card loading__multi-card\\"><div class=\\"content loading__content\\"><div class=\\"ui placeholder fluid\\"><div class=\\"square image\\"></div></div></div></div><div style=\\"height:20rem\\" class=\\"ui card loading__multi-card\\"><div class=\\"content loading__content\\"><div class=\\"ui placeholder fluid\\"><div class=\\"square image\\"></div></div></div></div><div style=\\"height:20rem\\" class=\\"ui card loading__multi-card\\"><div class=\\"content loading__content\\"><div class=\\"ui placeholder fluid\\"><div class=\\"square image\\"></div></div></div></div><div style=\\"height:20rem\\" class=\\"ui card loading__multi-card\\"><div class=\\"content loading__content\\"><div class=\\"ui placeholder fluid\\"><div class=\\"square image\\"></div></div></div></div><div style=\\"height:20rem\\" class=\\"ui card loading__multi-card\\"><div class=\\"content loading__content\\"><div class=\\"ui placeholder fluid\\"><div class=\\"square image\\"></div></div></div></div><div style=\\"height:20rem\\" class=\\"ui card loading__multi-card\\"><div class=\\"content loading__content\\"><div class=\\"ui placeholder fluid\\"><div class=\\"square image\\"></div></div></div></div></div></div></div><div class=\\"row\\"><div class=\\"stretched column\\"></div></div></div></div>"`;
